feat: water_probe_acquanativa_ap3 component definition

wvmcastro opened the pull request:
Depends on:
  - [ ]  https://github.com/tidewise/drivers-water_probe_acquanativa_ap3/pull/1

Definition of the water_probe_acquanativa_ap3 driver's orogen component. The component is a data source that pools the sensor every 0.1s and publishes the raw data into its `_data` output port.

wvmcastro committed:
feat: orogen component specification

wvmcastro committed:
feat: defining component's main task

DeboraTahara commented:
Change the title of this PR to clarify the main purpose of the PR, and use the tags 'fix, feat or chore' on the title and also in all commit messages. For example: 
feat: Implement the  water_probe_acquanativa component

DeboraTahara commented:
The format to use 'Depends on' syntax is:
Depends on:
- [ ] https://github.com/tidewise/drivers-water_probe_acquanativa_ap3/pull/1

wvmcastro committed:
feat: making this component a specialization of the iobase_drivers component

wvmcastro committed:
feat: following the iodrivers_base configureHook template

wvmcastro committed:
feat: overriding processIO method in order to pool sensor data

wvmcastro committed:
feat: remove unnecessary port connection check

wvmcastro committed:
feat: check for valid modbus address

wvmcastro committed:
feat: remove unnecessary port address check

this check is already being made by the base class

wvmcastro committed:
feat: add properties checking tests

wvmcastro committed:
chore: add modbus helpers from motors_weg_cvw300 driver orogen package

wvmcastro committed:
fix: remove io_port property setting test

Although it is necessary for the component's correct setup, it was removed in order to be possible to run tests

wvmcastro committed:
fix: remove sensor pooling from processIO to updateHook

wvmcastro committed:
feat: add modbus tests

kapeps commented on water_probe_acquanativa_ap3.orogen:
```diff
@@ -12,44 +14,27 @@ import_types_from "water_probe_acquanativa_ap3Types.hpp"
-    # in the C++ code with # _config_value.get() and _config_value.set(new_value).
-    property "config_value", "/std/string"
+    # Data output, each field reflects a mesurement
+    output_port "data", "water_probe_acquanativa_ap3/ProbeMeasurements"
```
Chose a better name for this output port

wvmcastro replied:
Do you have any suggestion? I did this way because later (when used in a composition) it will receive the suffix `_out` and will become `data_out`. Did I get that wrong?

kapeps replied:
No, "data" is just too general, normally we try to name the port to specify exactly what it is. Maybe "probe_measurements"

kapeps commented on test/task_test.rb:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        max_value = 2**16
+        modbus_set(registers["dissolved_oxygen"], rand(max_value))
+        modbus_set(registers["dissolved_oxygen_sat"], rand(max_value))
+        modbus_set(registers["temperature"], rand(max_value))
+        modbus_set(registers["ph"], rand(max_value))
+        modbus_set(registers["conductivity"], rand(max_value))
+        modbus_set(registers["salinity"], rand(max_value))
+        modbus_set(registers["dissolved_solids"], rand(max_value))
+        modbus_set(registers["specific_gravity"], rand(max_value))
+        modbus_set(registers["orp"], rand(max_value))
+        modbus_set(registers["turbity"], rand(max_value))
+        modbus_set(registers["height"], rand(max_value))
+        modbus_set(registers["latitude"], rand(max_value))
+        modbus_set(registers["longitude"], rand(max_value))
```
You should put chosen values and check for them in the "successfully read all measurements" test

rayssinha-soares replied:
but the library checks this values in https://github.com/tidewise/drivers-water_probe_acquanativa_ap3/ ... we thoght that the component does not write the values, it just requires data from the output port. Should we change it?

wvmcastro replied:
It is library's responsibility to check weather the data is right or not. Here I just check if given that there is data to be read, it is read by in the `updateHook`.

wvmcastro replied:
update: Talked with @doudou, and he also thinks that the component should check the values too. Maybe not as strictly as the lib, but at least some.

kapeps commented on test/task_test.rb:
```diff
@@ -0,0 +1,98 @@
+            modbus_configure_and_start
+        end
+
+        it "successfully read all measurements" do
+            modbus_expect_execution(@writer, @reader) do
+                mock_all_sensor_registers
+            end.to do
+                have_one_new_sample task.data_port
+            end
+        end
+
+        it "successfully read all measurements" do
+            modbus_expect_execution(@writer, @reader) do
+                mock_all_sensor_registers
+            end.to do
+                have_one_new_sample task.data_port
+            end
+        end
```
Two equal tests

kapeps commented on test/task_test.rb:
```diff
@@ -0,0 +1,98 @@
+
+        it "fail when measurements are not avaiable" do
+            modbus_expect_execution(@writer, @reader) do
+                have_no_new_sample task.data_port
```
you should specify the time(I think the default is 5 seconds, its too long), so it should be:
>have_no_new_sample task.data_port, at_least_during: 0.5

wvmcastro committed:
fix: remove duplicated test

doudou commented on test/task_test.rb:
```diff
@@ -0,0 +1,30 @@
+        it "cheks for device address setting obligatoriness" do
+            task.properties.io_port = "serial:///dev/ttyDUMMY:9600"
+            expect_execution.scheduler(true).to { fail_to_start task }
+        end
```
You must write a test that checks that you actually read data from the expected device, using the modbus helpers from motors_weg_cvw300

wvmcastro replied:
I thought I was doing that in the `"successfully read all measurements"` test. What I understood is that the modbus helpers emulate the device so I can read/write data from/to this emulated device. So, I created a fake bank of registers (using `mock_all_sensor_registers`) from which the component pools data from when its `updateHook` is called.

wvmcastro committed:
chore: comment justifying why processIO method is left empty

wvmcastro committed:
chore: set waiting sample timeout

wvmcastro committed:
chore: change output port name from data to probe_measurements

wvmcastro committed:
test: check for timestamp update

wvmcastro committed:
test: check driver raw data processing

wvmcastro committed:
fix(test): use same Time.now to calculate delta time

rayssinha-soares commented on test/task_test.rb:
```diff
@@ -0,0 +1,126 @@
+
+    def mock_all_sensor_registers
+        registers = {
+            "dissolved_oxygen" => 0,
+            "dissolved_oxygen_sat" => 1,
+            "temperature" => 2,
+            "ph" => 3,
+            "conductivity" => 4,
+            "salinity" => 5,
+            "dissolved_solids" => 6,
+            "specific_gravity" => 7,
+            "orp" => 8,
+            "turbity" => 9,
+            "height" => 10,
+            "latitude" => 11,
+            "longitude" => 12
```
```suggestion
            "oxygen_concentration" => 0,
            "oxygen_saturation" => 1,
            "temperature" => 2,
            "pH" => 3,
            "conductivity" => 4,
            "salinity" => 5,
            "dissolved_solids" => 6,
            "specific_gravity" => 7,
            "oxidation_reduction_potential" => 8,
            "turbidity" => 9,
            "height" => 10,
            "latitude" => 11,
            "longitude" => 12
```

rayssinha-soares commented on test/task_test.rb:
```diff
@@ -0,0 +1,126 @@
+        }
+
+        @probe_measurements_raw = [756, 85, 200, 646, 156, 126, 1256, 32, 40, 3, 500, 1, 76]
+        modbus_set(registers["dissolved_oxygen"], @probe_measurements_raw[0])
+        modbus_set(registers["dissolved_oxygen_sat"], @probe_measurements_raw[1])
+        modbus_set(registers["temperature"], @probe_measurements_raw[2])
+        modbus_set(registers["ph"], @probe_measurements_raw[3])
+        modbus_set(registers["conductivity"], @probe_measurements_raw[4])
+        modbus_set(registers["salinity"], @probe_measurements_raw[5])
+        modbus_set(registers["dissolved_solids"], @probe_measurements_raw[6])
+        modbus_set(registers["specific_gravity"], @probe_measurements_raw[7])
+        modbus_set(registers["orp"], @probe_measurements_raw[8])
+        modbus_set(registers["turbity"], @probe_measurements_raw[9])
```
```suggestion
        modbus_set(registers["oxygen_concentration"], @probe_measurements_raw[0])
        modbus_set(registers["oxygen_saturation"], @probe_measurements_raw[1])
        modbus_set(registers["temperature"], @probe_measurements_raw[2])
        modbus_set(registers["pH"], @probe_measurements_raw[3])
        modbus_set(registers["conductivity"], @probe_measurements_raw[4])
        modbus_set(registers["salinity"], @probe_measurements_raw[5])
        modbus_set(registers["dissolved_solids"], @probe_measurements_raw[6])
        modbus_set(registers["specific_gravity"], @probe_measurements_raw[7])
        modbus_set(registers["oxidation_reduction_potential"], @probe_measurements_raw[8])
        modbus_set(registers["turbidity"], @probe_measurements_raw[9])
```

rayssinha-soares commented on test/task_test.rb:
```diff
@@ -0,0 +1,126 @@
+
+    def get_expected_sample
+        sample = Types.water_probe_acquanativa_ap3.ProbeMeasurements.new
+        sample.concentration = @probe_measurements_raw[0] * 1e-5
+        sample.saturation = @probe_measurements_raw[1] * 1e-2
+        sample.temperature.kelvin = 275.15;
+        sample.ph = @probe_measurements_raw[3] * 1e-2;
+        sample.conductivity = @probe_measurements_raw[4] * 1e-10;
+        sample.salinity = @probe_measurements_raw[5] * 1e-2
+        sample.dissolved_solids = @probe_measurements_raw[6] * 1e-2
+        sample.specific_gravity = @probe_measurements_raw[7] * 1e-2
+        sample.ORP = @probe_measurements_raw[8] * 1e-3
+        sample.turbity = @probe_measurements_raw[9]
```
```suggestion
        sample.oxygen_concentration = @probe_measurements_raw[0] * 1e-8
        sample.oxygen_saturation = @probe_measurements_raw[1] * 1e-4
        sample.temperature.kelvin = 275.15;
        sample.pH = @probe_measurements_raw[3] * 1e-2;
        sample.conductivity = @probe_measurements_raw[4] * 1e-10;
        sample.salinity = @probe_measurements_raw[5] * 1e-2
        sample.dissolved_solids = @probe_measurements_raw[6] * 1e-8
        sample.specific_gravity = @probe_measurements_raw[7] * 1e-2
        sample.oxidation_reduction_potential = @probe_measurements_raw[8] * 1e-3
        sample.turbidity = @probe_measurements_raw[9]
```

wvmcastro committed:
Update test/task_test.rb

Co-authored-by: rayssinha-soares <79292544+rayssinha-soares@users.noreply.github.com>

wvmcastro committed:
Update test/task_test.rb

Co-authored-by: rayssinha-soares <79292544+rayssinha-soares@users.noreply.github.com>

wvmcastro committed:
Update test/task_test.rb

Co-authored-by: rayssinha-soares <79292544+rayssinha-soares@users.noreply.github.com>
